ignite/templates/ibc/packet.go

```diff
@@ -171,76 +174,84 @@ func moduleModify(replacer placeholder.Replacer, opts *PacketOptions) genny.RunF
 	}
 }
 
-func protoModify(replacer placeholder.Replacer, opts *PacketOptions) genny.RunFn {
+// Modifies packet.proto to add a field on the oneof element of the message created and
+// add a couple of messages.
+//
+// What it depends on:
+//   - Existence of a Oneof field named 'packet'.
+func protoModify(opts *PacketOptions) genny.RunFn {
 	return func(r *genny.Runner) error {
 		path := filepath.Join(opts.AppPath, "proto", opts.AppName, opts.ModuleName, "packet.proto")
 		f, err := r.Disk.Find(path)
 		if err != nil {
 			return err
 		}
-
-		content := f.String()
-
-		// Add the field in the module packet
-		fieldCount := strings.Count(content, PlaceholderIBCPacketProtoFieldNumber)
-		templateField := `%[1]v
-				%[2]vPacketData %[3]vPacket = %[4]v; %[5]v`
-		replacementField := fmt.Sprintf(
-			templateField,
-			PlaceholderIBCPacketProtoField,
-			opts.PacketName.UpperCamel,
-			opts.PacketName.LowerCamel,
-			fieldCount+2,
-			PlaceholderIBCPacketProtoFieldNumber,
-		)
-		content = replacer.Replace(content, PlaceholderIBCPacketProtoField, replacementField)
+		pf, err := protoutil.ParseProtoFile(f)
+		if err != nil {
+			return err
+		}
+		name := cases.Title(language.English, cases.Compact).String(opts.ModuleName) + "PacketData"
+		m, err := protoutil.GetMessageByName(pf, name)
+		if err != nil {
+			return fmt.Errorf("failed while looking up '%s' in %s: %w", name, path, err)
+		}
+		// Use a directly Apply call here, modifying oneofs isn't common enough to warrant a separate function.
+		var packet *proto.Oneof
+		protoutil.Apply(m, nil, func(c *protoutil.Cursor) bool {
+			if o, ok := c.Node().(*proto.Oneof); ok {
+				if o.Name == "packet" {
+					packet = o
+					return false
+				}
+			}
+			// continue traversing.
+			return true
+		})
+		if packet == nil {
+			return fmt.Errorf("could not find 'oneof packet' in %s", path)
+		}
+		// Count fields of oneof:
+		max := 1
+		protoutil.Apply(packet, nil, func(c *protoutil.Cursor) bool {
+			if o, ok := c.Node().(*proto.OneOfField); ok {
+				if o.Sequence > max {
+					max = o.Sequence
+				}
+			}
+			return true
+		})
```

Comment:
Genuine question, isn't it possible to get the number of fields in oneof from `len(packet.Elements)` ?

(I assume that `Elements` are the children, but maybe I'm wrong!)

Reply:
tested and not possible because `Elements` aren't necessarily of type `OneOfField`. I close this comment.

Reply:
Yeah, iirc its probably comment nodes.